### PR TITLE
Refactoring of log.cfg() and box.cfg() from the POV of log configuration

### LINKED ISSUE #7447: log.cfg panic on incorrect syslog config
```
Tarantool 2.11.0-entrypoint-210-g111663821
type 'help' for interactive help
 tarantool> log = require('log')
---
...

 tarantool> log.cfg{log='syslog:xxx'}
IllegalParams: bad option 'xxx'
failed to initialize logging subsystem
```
At the same time:
```
Tarantool 2.11.0-entrypoint-210-g111663821
type 'help' for interactive help
 tarantool> box.cfg{log='syslog:xxx'}
---
- error: 'Incorrect value for option ''log'': bad option ''xxx'''
...

 tarantool>
```
I guess we don't need to panic in the the first case as well.

### PR DESCRIPTION
**Pull request by nshy**
- Log configuration on first configuration thru box.cfg() is really painful:

    1. check log cfg validity in scope of private.cfg_check (1)
    2. set log option values to log.cfg in update_module_cfg. We also check log cfg validity (2) here in api.cfg_set and have rollback logic (why he have to?)
    3. make actual log initialization in private.cfg_load()
    4. update dynamic options of log cfg once again. Why? Probably only to make conversions like for log_level from 'warn' to number. We make cfg validity checks here too! (3)

I guess we'd better have only one validity check and don't run step 4 which actually reconfigure log once again only to perform conversion.

- Log API is painfully bloated:

    - log.level - to set log level by user
    - log.box_api.cfg_get - private API to just get log level (why not use just log.cfg.level? Well cfg_get works for 'log_level', a bit different name!)
    - log.box_api.cfg_set - private API to just SAVE log level (do not change log level actually, only saves values to log.cfg)
    - log.box_api.cfg_set_log_level - really change log level, the difference from log.level is it does not get value as argument but reads it from global box.cfg.log_level and throws box.error() instead of error())

And the same for log_format!

I propose to drop box_api completely and use log.cfg (function/table) from load_cfg.lua.

The PR does not fix multiple log config validity check issue completly. I do not include all the patches to keep PR commits number small.

Closes #7447

### CODE REVIEW

File: src/main.cc
Comment (locker) [inline src/main.cc:440]: It's important not to print anything to the other instance's log in this case so this comment was here for a reason. Please don't change this behavior.

**Review by locker — COMMENTED** [review]
General note: too many patches. I consider splitting a PR or squashing some patches if there are more than 10. IMO a PR should have no more than about 5 patches to look reviewer friendly. With 19 patches, I got lost where this PR is going somewhere at patch 8. As a result, I had to look through all the patches squashed together :-/

Please see a few comments inline.

File: src/box/lua/load_cfg.lua
Comment (cyrillos) [reply src/box/lua/load_cfg.lua:258]: It came from the request where `log` module need to be used _before_ `box.cfg` call. IOW customers wanted to setup log module, use it for some time and only later call `box.cfg`. Due to this I've had to implement a way more complex mappings as it should be (simply because logging never had a proper architecture but been bound to box.cfg).

File: src/box/box.cc
Comment (nshy) [reply src/box/box.cc:770]: Now the same code throws errors for box.cfg() and log.cfg(), so I thought `log_level` in the message can be unexpected for log.cfg() call. And `log level` is somewhat suits for both.

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua:258]: @locker Well it makes sense to me. This works while log module is configured only as part of the box configuration. Otherwise log (and other modules) have to keep adapters for every user. So typically approach when user cares of adapters looks more managable.

Anyway I can keep the original approach and still substantially reduce complexity.

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua:334]: This is not really so. `log_cfg.lua` has the list of dynamic options and so` log.lua` does. So this list is not that functional.

File: src/lua/log.lua
Comment (nshy) [reply src/lua/log.lua:435]: No, it is not global. I put `local log` at the beginning so that I can use `log.debug(...)` instead of `say(S_DEBUG, ...)`.

File: src/box/lua/load_cfg.lua
Comment (locker) [reply src/box/lua/load_cfg.lua:334]: Yep, but before this change, I could figure out which options could be changed dynamically by looking at just one table (`dynamic_cfg`). Now, it isn't so - I have to remember that some options are updated directly by `reload_cfg`. Not critical, but a bit inconvenient.

File: src/box/lua/load_cfg.lua
Comment (locker) [reply src/box/lua/load_cfg.lua:258]: Overall, I'm not sure which approach is better. The one thing that I liked about the old approach is that it kept `load_cfg` infrastructure simple by offloading all the complexity to modules. There were more code, but interaction between different subsystems were clearer IMO. I wouldn't bother reworking it unless it's necessary to fix a bug or implement a new feature. I won't block it, but I won't approve it, either, unless you get an approval from a second reviewer from our team, preferably, from @alyapunov.

Comment (nshy) [reply src/box/lua/load_cfg.lua:258]: Ok. I'm making some changes after the review now. So let's check second version.

**Comment by nshy** [comment]
Beside making code nicer this PR will help to fix #7447

File: src/box/lua/load_cfg.lua
Comment (locker) [inline src/box/lua/load_cfg.lua:772]: Why do you need this `pcall`? You re-throw an error if it fails.

File: src/lua/log.lua
Comment (locker) [inline src/lua/log.lua:419]: There's a typo in this code:

```
tarantool> box.error(box.error.cfg, 'log nonblock', "must be 'true' or 'false'")
---
- error: 'box.error(): bad arguments'
...
```

Apparently, we lack tests that check errors raised on invalid log configuration :-(

File: src/lua/log.lua
Comment (locker) [inline src/lua/log.lua:437]: I don't like that now we use `ER_CFG`, which was original thrown only on `box.cfg` errors, in the log module. Aesthetically, this doesn't look good to me, but it may be all right. @alyapunov WDYT?

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua:778]: module.check expects parameters with module naming. For example `level` and not `log_level`.

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua:657]: There is no rollback here really. In case of error we only try to rollback failed option. And module.cfg is atomic.

File: src/lua/log.lua
Comment (nshy) [reply src/lua/log.lua:437]: In terms of PR this allows to reduce code. And in terms of meaning this is just same configuration error just like in case of box.

Comment (nshy) [reply src/lua/log.lua:437]: It will be really painful to throw box.error or error depending on calling context.

File: src/box/lua/load_cfg.lua
Comment (locker) [reply src/box/lua/load_cfg.lua:778]: Can we use the same names to avoid remapping?

File: src/box/lua/load_cfg.lua
Comment (locker) [reply src/box/lua/load_cfg.lua:657]: You're right.

I found an issue with this code:

```
tarantool> box.cfg{log_level = 'warn'}
---
- error: Can't set option 'log_level' dynamically
...
```

Before this patch, one could change the option dynamically. I think we need to add some tests for logger configuration before applying this patch.

**Comment by alyapunov** [comment]
Commits that are pushed to master must reference some issue. While the patch is under review perhaps it's OK, but there must be at least reference in PR. Or how can I understand which task are you solving?

**Review by Gumix — APPROVED** [review]

File: src/box/box.cc
Comment (locker) [inline src/box/box.cc:836]: `__setter` pattern is confusing. Please split into two functions instead.

File: src/box/lua/load_cfg.lua
Comment (locker) [inline src/box/lua/load_cfg.lua:404]: Bad naming: "settings" is a synonym to "options", which looks confusing. Discussed f2f and agreed that `log.cfg` should update `box.internal.cfg` directly, which will be then used on first call to `box.cfg`.

File: src/box/lua/load_cfg.lua
Comment (locker) [inline src/box/lua/load_cfg.lua:410]: Please remove the name mapping, because it's inconsistent with other modules and complicates the API. The `box_api.cfg` function should do the mapping by itself.

File: src/box/lua/load_cfg.lua
Comment (locker) [inline src/box/lua/load_cfg.lua:405]: There's a bug somewhere in this code. The following commands work although they shouldn't:

```
tarantool> require('log').cfg{log = 'foo.log'}
---
...

tarantool> box.cfg{log = 'bar.log'}
---
...
```

Please fix and add a test case.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/box.cc
Comment (nshy) [reply src/box/box.cc]: Not applicable in new version.

File: src/box/box.cc
Comment (nshy) [reply src/box/box.cc]: Fixed.

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua]: Fixed.

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua]: Fixed.

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua]: Bug is fixed and test is added.

File: src/main.cc
Comment (locker) [inline src/main.cc:481]: Nit: Why `panic` here? Why not `exit` as in other places? Looks inconsistent.

File: src/box/log.h
Comment (locker) [inline src/box/log.h:15]: This function declaration is redundant - we never call it from C code. Anyway, I don't think that this function deserves a separate file - let's leave it in `box.cc`, where we have all other config-checking functions? Or maybe move it to `src/lib/core/say.c` and return config name + error string instead of `ER_CFG`? (The core lib can't use `ClientError`.) Or maybe rewrite it completely in Lua?

File: src/lua/log.lua
Comment (locker) [inline src/lua/log.lua]: I found it confusing at first that this patch leaves the old methods be although it looks like they are never used. I now realize that they are removed in the next patch. Why not squash the patches to avoid confusion?

File: src/lua/log.lua
Comment (locker) [inline src/lua/log.lua:122]: Nit: Looks like this `log` forward declaration is required solely to use `log.debug` in the functions below. If so, it'd be better to define `log_debug` local function instead.

File: src/lua/log.lua
Comment (locker) [inline src/lua/log.lua:274]: Nit: It'd be nice to have brief comments for this function and all the new helper functions below. They aren't quite trivial so comments would help.

File: src/box/log.h
Comment (locker) [inline src/box/log.h:22]: The `extern "C"` header is obviously misplaced.

File: src/lua/log.lua
Comment (locker) [inline src/lua/log.lua:335]: Nit: You don't need to use `string.format` here. This would work as well:

```Lua
log.debug("log: level set to %s", level)
```

File: src/lua/log.lua
Comment (locker) [inline src/lua/log.lua:355]: Looks like `prepare_cfg` and `merge_cfg` are always used together (in `load_cfg.lua` too). Maybe, merge them to make the internal API simpler?

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua]: Done.

File: src/box/lua/load_cfg.lua
Comment (nshy) [reply src/box/lua/load_cfg.lua:1000]: Done.

File: src/lib/core/say.c
Comment (nshy) [reply src/lib/core/say.c]: Fixed.

File: src/main.cc
Comment (nshy) [reply src/main.cc]: Fixed.

File: src/box/log.h
Comment (nshy) [reply src/box/log.h]: As discussed offline - leave it in box.cc.

File: src/lua/log.lua
Comment (nshy) [reply src/lua/log.lua]: Done.

File: src/lua/log.lua
Comment (nshy) [reply src/lua/log.lua:280]: Done.

File: src/box/log.h
Comment (nshy) [reply src/box/log.h]: Dropped log.h file.

File: src/lua/log.lua
Comment (nshy) [reply src/lua/log.lua]: Fixed.

File: extra/dist/tarantoolctl.in
Comment (nshy) [reply extra/dist/tarantoolctl.in:457]: Reverted changes for this this part.

File: src/box/box.cc
Comment (nshy) [reply src/box/box.cc:764]: Yep, I mistakenly thought we miss cleanup in case of error. Fixed.

File: src/box/box.cc
Comment (nshy) [reply src/box/box.cc:1420]: Done.

File: src/lib/core/say.c
Comment (nshy) [reply src/lib/core/say.c:769]: Next time)

**Review by locker — APPROVED** [review]

**Review by Gumix — APPROVED** [review]

**Review by alyapunov — APPROVED** [review]

**Comment by locker** [comment]
@nshy please rebase and resolve conflicts.

**Pull request merged**
